File: src/expr/like.rs
// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use regex::Regex;

pub fn build_like_regex_from_string(like_string: &str) -> Result<Regex, failure::Error> {
    // The goal is to build a regex that matches the same strings as the LIKE
    // pattern.
    //
    // LIKE patterns always cover the whole string, so we anchor the regex on
    // both sides. An underscore (`_`) in a LIKE pattern matches any single
    // character and a percent sign (`%`) matches any sequence of zero or more
    // characters, so we translate those operators to their equivalent regex
    // operators, `.` and `.*`, respectively. Other characters match themselves
    // and are copied directly, unless they have special meaning in a regex, in
    // which case they are escaped in the regex with a backslash (`\`).
    //
    // Note that characters in LIKE patterns may also be escaped by preceding
    // them with a backslash. This has no effect on most characters, but it
    // removes the special meaning from the underscore and percent sign
    // operators, and means that matching a literal backslash requires doubling
    // the backslash.
    //
    // TODO(benesch): SQL permits selecting a different escape character than
    // the backslash via LIKE '...' ESCAPE '...'. We will need to support this
    // syntax eventually.
    let mut regex = String::from("^");
    let mut escape = false;
    for c in like_string.chars() {
        match c {
            '\\' if !escape => escape = true,
            '_' if !escape => regex.push('.'),
            '%' if !escape => regex.push_str(".*"),
            c => {
                if regex_syntax::is_meta_character(c) {
                    regex.push('\\');
                }
                regex.push(c);
                escape = false;
            }
        }
    }
    regex.push('$');
    if escape {
        failure::bail!("unterminated escape sequence in LIKE")
    } else {
        Ok(Regex::new(&regex)?)
    }
}

File: misc/python/materialize/checks/like.py
# Copyright Materialize, Inc. and contributors. All rights reserved.
#
# Use of this software is governed by the Business Source License
# included in the LICENSE file at the root of this repository.
#
# As of the Change Date specified in that file, in accordance with
# the Business Source License, use of this software will be governed
# by the Apache License, Version 2.0.
from textwrap import dedent

from materialize.checks.actions import Testdrive
from materialize.checks.checks import Check


class Like(Check):
    """LIKE, ILIKE, with and without a constant pattern are all compiled and evaluated differently"""

    def initialize(self) -> Testdrive:
        return Testdrive(
            dedent(
                """
            > CREATE TABLE like_regex_table (f1 STRING, f2 STRING, f3 STRING, f4 STRING);
            > INSERT INTO like_regex_table VALUES ('abc', 'abc', 'a~%', '~');
            """
            )
        )

    def manipulate(self) -> list[Testdrive]:
        return [
            Testdrive(dedent(s))
            for s in [
                """
                > CREATE MATERIALIZED VIEW like_regex_view1 AS SELECT f1 LIKE f2 AS c1, f1 ILIKE f2 AS c2, f1 LIKE 'x_z' AS c3, f1 ILIKE 'a_c' AS c4, f1 LIKE 'a~%' ESCAPE '~' AS c5, f1 LIKE f3 ESCAPE '~' AS c6, f1 LIKE f3 ESCAPE f4 AS c7 FROM like_regex_table;
                > INSERT INTO like_regex_table VALUES ('klm', 'klm', 'k_m', 'k');
            """,
                """
                > CREATE MATERIALIZED VIEW like_regex_view2 AS SELECT f1 LIKE f2 AS c1, f1 ILIKE f2 AS c2, f1 LIKE 'x_z' AS c3, f1 ILIKE 'a_c' AS c4, f1 LIKE 'a~%' ESCAPE '~' AS c5, f1 LIKE f3 ESCAPE '~' AS c6, f1 LIKE f3 ESCAPE f4 AS c7 FROM like_regex_table;
                > INSERT INTO like_regex_table VALUES ('xyz', 'xyz', 'x_%', '_');
            """,
            ]
        ]

    def validate(self) -> Testdrive:
        return Testdrive(
            dedent(
                """
            > SELECT * FROM like_regex_view1;
            true true false true false false false
            true true false false false true false
            true true true false false true false

            > SELECT * FROM like_regex_view2;
            true true false true false false false
            true true false false false true false
            true true true false false true false
            """
            )
        )
